### ISSUE TITLE
quoted double-quotes can wrongly trigger "open ended" compilation warnings

### ISSUE THREAD

**Issue by aspiers** [supercollider/supercollider#3352]
This bug is very similar to #251, and I'm using 3.8.0 (on openSUSE Leap 42.2, FWIW), so it's possible it's been fixed in a later version.

Placing this class somewhere in the compile path:

```
OpenEndedStringTestCase {
	var foo = "\"";
	var bar = "bar";
}
```

will result in a compile-time warning like:

    Open ended string ... started on line 3 in file '/home/adam/music/software/SuperCollider/quarks/MyTests/doublequote_test.sc'

A workaround is to append a comment to the line with the escaped double-quote:

	var foo = "\"";  // " to trick the interpreter into "closing" the quote

If `"bar"` is changed to `"'"`, the warning changes to `Open ended symbol ...`.

If this is indeed still present in latest `git`, I suspect that caution will be required in order to avoid regressing #251 when fixing it.

**Comment by snappizz** [comment]
i'm not seeing this in latest 3.9 on ubuntu. what OS are you using?

**Comment by brianlheim** [comment]
Cannot reproduce this on macOS.

An initial guess is that there's UB in the lexer from an uninitialized value or dangling pointer.

**Comment by aspiers** [comment]
I updated my report to mention my OS.  Sounds like maybe it's been fixed since 3.8.0.

**Comment by aspiers** [comment]
I'm building from git now to find out.

**Comment by aspiers** [comment]
I have just reproduced this with latest git `master` on my openSUSE Leap 42.2 system.

**Comment by brianlheim** [comment]
@aspiers the main branch is now develop - master hasn't been updated in quite awhile. Sorry for the confusion.

**Comment by aspiers** [comment]
OK, trying with `develop`...

**Comment by aspiers** [comment]
Ah no, I was already on `develop` without noticing, since that's the default branch on GitHub.  So this is reproducible on `develop`.

**Comment by aspiers** [comment]
Having said that, it looks like `3.9` is a slightly newer branch ...

**Comment by patrickdupuis** [comment]
Yes, please try to test this on the 3.9 branch. We are a few days away from our next beta release of SC 3.9.

**Comment by aspiers** [comment]
I just reproduced with the `3.9` branch.

**Comment by snappizz** [comment]
well that is bizarre... @brianlheim i wonder if we should just tear down `PyrLexer` and redo it from the ground up.

**Comment by brianlheim** [comment]
@aspiers - with your workaround, what is the observed value of `foo` in the example above?

**Comment by brianlheim** [comment]
Lol I can reproduce this now while debugging. Investigating...

**Comment by brianlheim** [comment]
OK, I think I found it. Also, just to be clear (it wasn't to me originally) this is a spurious warning but doesn't actually halt compilation. The value of `foo` for me is `"` as expected, and I would guess it is for you too. If not, that makes this situation worse.

Please try this diff: 

```diff
diff --git a/lang/LangSource/PyrLexer.cpp b/lang/LangSource/PyrLexer.cpp
index 42da0b1b2..6fadb565b 100644
--- a/lang/LangSource/PyrLexer.cpp
+++ b/lang/LangSource/PyrLexer.cpp
@@ -1375,7 +1375,7 @@ symbol3 : {
                do {
                        c = input0();
                        if (c == '\\') {
-                               c = input0();
+                               input0();
                        }
                } while (c != endchar && c != 0);
                if (c == 0) {
@@ -1394,7 +1394,7 @@ string1 : {
                do  {
                        c = input0();
                        if (c == '\\') {
-                               c = input0();
+                               input0();
                        }
                } while (c != endchar && c != 0);
                if (c == 0) {
```

I can't tell why the behavior on this changes in the optimizer.

>  @brianlheim i wonder if we should just tear down PyrLexer and redo it from the ground up.

I think it can be sanely refactored now that we have regression tests. Most of the code is accurate and runs about as fast as it should. The main obstacle is that it isn't written testably. Overall I agree. This code produces a disproportionate number of bugs, it deserves more attention.

**Comment by jamshark70** [comment]
Another example:

```supercollider
TestStringParser {
	*initClass {
		"\"".postln;
		("\"".postln);
	}
}
```

**Comment by patrickdupuis** [comment]
Fixed in #4255
